RavenDB-25416 - Reduce TopDocs memory footprint using ManagedScoreDocArray

grisha-kotler opened the pull request:
**Replace ScoreDoc[] with Zero-Allocation ManagedScoreDocArray**

**The Problem**
In high-scale search scenarios returning large result sets (e.g., millions of hits), the legacy `ScoreDoc[]` implementation creates significant memory and GC bottlenecks:

1.  **Excessive Memory Overhead:**
      * Each `ScoreDoc` is a class, incurring a 24-byte heap overhead per result.
      * Combined with the array pointer, each hit costs **32 bytes**.
      * *Example:* 10 million hits $\approx$ **320 MB** of active heap memory.
2.  **LOH Fragmentation:**
      * The `ScoreDoc[]` array itself exceeds the 85KB Large Object Heap (LOH) threshold for any result set \> 10,600 items. This causes heap fragmentation and expensive Gen 2 collections.
3.  **GC Pressure & Latency:**
      * Allocating millions of tiny, short-lived `ScoreDoc` objects floods Gen 0/Gen 1, forcing frequent Garbage Collection cycles.
      * Creates a very large reference graph which increases the GC time (https://issues.hibernatingrhinos.com/issue/RavenDB-24815/Large-reference-graph-for-arrays-of-strings).
4.  **Poor Data Locality:**
      * `ScoreDoc[]` is an array of pointers. Iterating it requires "pointer chasing" (jumping to random heap locations), which destroys CPU cache efficiency.

**The Solution: ManagedScoreDocArray**
Replaced the array of objects with a struct-of-arrays pattern (`int[] docs`, `float[] scores`) managed via `ArrayPool`.

1.  **4x Memory Reduction:**
      * We now store only raw primitives: 4 bytes (`int`) + 4 bytes (`float`) = **8 bytes per hit**.
      * *Example:* 10 million hits $\approx$ **80 MB** (down from 320 MB).
2.  **Zero-Allocation & LOH Avoidance:**
      * Buffers are rented from `ArrayPool<T>`, eliminating allocation churn.
      * Internal segments are capped at 64KB, ensuring they **never** enter the LOH.
3.  **Cache-Friendly:**
      * Data is stored contiguously. Scans and sorts benefit from CPU prefetching and SIMD optimizations.

The `TopDocs` class retains the `ScoreDoc[]` property but marks it as `[Obsolete]`. Accessing it will lazily materialize the old array format for compatibility which is only used in tests, but the internal engine uses the efficient `ManagedScoreDocArray`.

Comparison of the peak working set during the execution of the tests.
| Implementation | Peak Working Set | Reduction vs Old |
| :--- | :--- | :--- |
| **Baseline** | 3.967 GB | - |
| 🏆 **ArrayPool** | **920 MB** | **~4.3x less** |

What does this "Memory Reduction" actually mean?  We are specifically talking about the managed memory allocated during the execution of array read and write operations.
In the old implementation, these hot paths generated a massive amount of short-lived objects. Even though most of these allocations were trapped in Gen 0, the Garbage Collector (GC) still had to constantly burn CPU cycles to track and clean them up.

The following benchmarks compare the previous approach (using a List materialized to an array) against the new approach. Both tests were conducted after a warmup phase. In the legacy implementation, this ensured the list returned to the object pool for reuse; the new implementation utilizes ArrayPool. Note that the reported **execution time** is **per request**, while the **allocation** represents the **total memory allocated** throughout the entire process.

* **Peak Speedup**: The highest speedup occurs at maximum load (131k docs, 32 tasks), where the new system is nearly **38x faster**.
* **Peak Efficiency**: Memory reduction is most dramatic in larger datasets, consistently reducing allocations by **500x to 700x**.
* **Scaling**: As the number of tasks increases, the performance gap widens significantly in favor of the new system.

| Docs | Tasks | Old Time (ms) | New Time (ms) | Speedup | Old Alloc | New Alloc | Memory Reduction |
| :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- |
| 128 | 1 | 0.00240 | 0.00117 | **2.1x** | 515 KB | 22 KB | **23x** |
| 128 | 4 | 0.00163 | 0.00069 | **2.4x** | 2.01 MB | 88 KB | **23x** |
| 128 | 8 | 0.00112 | 0.00067 | **1.7x** | 4.02 MB | 176 KB | **23x** |
| 1,024 | 1 | 0.00184 | 0.00056 | **3.3x** | 4.02 MB | 51.07 KB | **81x** |
| 16,384 | 4 | 0.00067 | 0.00018 | **3.7x** | 257.18 MB | 608.33 KB | **433x** |
| 16,384 | 8 | 0.00169 | 0.00030 | **5.6x** | 513.20 MB | 1.16 MB | **442x** |
| 16,384 | 16 | 0.00101 | 0.00008 | **12.6x** | 1.00 GB | 1.93 MB | **531x** |
| 16,384 | 32 | 0.00091 | 0.00007 | **13.0x** | 2.00 GB | 2.99 MB | **685x** |
| 131,072 | 1 | 0.00241 | 0.00038 | **6.3x** | 512.00 MB | 255.14 KB | **2,055x** |
| 131,072 | 4 | 0.00125 | 0.00013 | **9.6x** | 2.00 GB | 1.12 MB | **1,829x** |
| 131,072 | 8 | 0.00116 | 0.00010 | **11.6x** | 4.01 GB | 2.05 MB | **2,003x** |
| 131,072 | 16 | 0.00108 | 0.00006 | **18.0x** | 8.02 GB | 10.05 MB | **817x** |
| 131,072 | 32 | 0.00106 | 0.00006 | **17.7x** | 16.03 GB | 8.86 MB | **1,853x** |

grisha-kotler committed:
RavenDB-25416 - add the ManagedScoreDocArray

grisha-kotler committed:
RavenDB-25416 - fix failing tests

grisha-kotler committed:
RavenDB-25416 - mark materializing the ScoreDocs as obsolete

grisha-kotler committed:
RavenDB-25416 - add the ManagedScoreDocArray tests

grisha-kotler committed:
RavenDB-25416 - apply Unsafe ref arithmetic optimization

grisha-kotler committed:
RavenDB-25416 - remove the GetRawValues

grisha-kotler committed:
RavenDB-25416 - apply the same optimization for the fields

grisha-kotler committed:
RavenDB-25416 - cleanup

grisha-kotler committed:
RavenDB-25416 - add a test for fields

maciejaszyk commented on src/Lucene.Net/Util/ManagedScoreDocArray.cs:
```diff
@@ -0,0 +1,537 @@
+        ref float scoresRef = ref MemoryMarshal.GetArrayDataReference(_currentScores);
+
+        Unsafe.Add(ref docsRef, _currentSegmentUsed) = doc;
+        Unsafe.Add(ref scoresRef, _currentSegmentUsed) = score;
```
Does it make sense to do this for just one assignment per array? I remember a case from Sparrow where `MemoryMarshal.GetReference` was very costly.

grisha-kotler replied:
done

maciejaszyk commented on src/Lucene.Net/Util/ManagedScoreDocArray.cs:
```diff
@@ -0,0 +1,537 @@
+    private const int GrowthPhaseTotalItems = 16128;
+
+    // there are 6 segments in the growth phase (0 to 5)
+    // segment 6 is the first "Stable" (max Size) segment.
```
Can you elaborate more (and put in as a comment) about that phases etc?

grisha-kotler replied:
done

arekpalinski commented:
> The following benchmarks compare ... 

Can you please share this benchmark code?

maciejaszyk commented on src/Lucene.Net/Util/ManagedScoreDocArray.cs:
```diff
@@ -0,0 +1,537 @@
+        return new BackwardsWriter(this);
+    }
+
+    public struct BackwardsWriter
```
I think you might want to make this a ref struct. I see it lives only on the stack, so it gives you the opportunity to create local ref variables like:
```
        private ref int _currentDocsRef;
        private ref float _currentScoreRef;
        private ref IComparable[] _currentFieldsRef;
```

and avoid retrieving 
```
ref int docsRef = ref MemoryMarshal.GetArrayDataReference(_currentDocs);
ref float scoresRef = ref MemoryMarshal.GetArrayDataReference(_currentScores);
```
in `Write` method. Just load it once per segment.

grisha-kotler replied:
done

grisha-kotler replied:
I also ran the benchmark again.

maciejaszyk commented on src/Lucene.Net/Util/ManagedScoreDocArray.cs:
```diff
@@ -0,0 +1,537 @@
+        }
+
+        var docs = ArrayPool<int>.Shared.Rent(newSize);
+        var scores = ArrayPool<float>.Shared.Rent(newSize);
```
We had performance issues with shared ArrayPools during Corax's development.
You might be interested:
https://github.com/ravendb/ravendb/blob/d4fe36eb10bad4b0d379d0efedfba843e2cbf172/src/Corax/Analyzers/Analyzer.cs#L17-L23

Scooletz replied:
What @maciejaszyk said above + some links to the runtime repo: 

[SharedArrayPool](https://github.com/dotnet/runtime/blob/cbabb9aca41669a5755d2f81d23b5fe56c3aaf6c/src/libraries/System.Private.CoreLib/src/System/Buffers/SharedArrayPool.cs#L50-L126) tries to get a thread-local buffer first. When it fails, it moves to shared partitions. The [partition though is synchronized](https://github.com/dotnet/runtime/blob/cbabb9aca41669a5755d2f81d23b5fe56c3aaf6c/src/libraries/System.Private.CoreLib/src/System/Buffers/SharedArrayPool.cs#L426-L435) so may run into perf degradation if you borrow a lot and return a lot. As usual, worth to measure :-)

grisha-kotler replied:
Isn't using `ThreadStatic` a waste?
Shouldn't it use an `ArrayPool` per core?

Scooletz replied:
As discussed in a meeting, I think we could start with the shared being extracted to a static field. Being given the segments being allocated in doubling sized, we might hit the sweet spot of the shared one. If this does not work, we can switch to some other pooling, either per core or ThreadStatic.

grisha-kotler replied:
done.

grisha-kotler commented:
> > The following benchmarks compare ...
> 
> Can you please share this benchmark code?

Benchmark code was added here: https://issues.hibernatingrhinos.com/issue/RavenDB-25416/High-memory-usage-during-streaming-query#focus=Comments-67-1071631.0-0

grisha-kotler committed:
RavenDB-25416 - skip collecting the fields for sorting when not needed

grisha-kotler committed:
RavenDB-25416 - cleanup

grisha-kotler committed:
RavenDB-25416 - bump lucene version

grisha-kotler committed:
RavenDB-25416 - Refactored ManagedScoreDocArray to pack doc (int) and score (float) into a single long[] array instead of separate arrays to improve memory locality and reduce object allocations.
Updated segment capacities and implemented Unsafe memory manipulation to efficiently read and write the packed 8-byte structures while staying safely under the LOH threshold.

grisha-kotler committed:
RavenDB-25416 - never dispose the static Empty instance

grisha-kotler committed:
RavenDB-25416 - bump lucene version

Scooletz commented on src/Lucene.Net/Util/ManagedScoreDocArray.cs:
```diff
@@ -0,0 +1,509 @@
+
+    private void AllocateSegment(int size, bool hasFields)
+    {
+        var packed = ArrayPool<long>.Shared.Rent(size);
```
I'd consider accepting and `ArrayPool<long>` in the ctor of this class, then use it here and in the disposal. We should not be making assumptions on such low level.

grisha-kotler replied:
done

Scooletz commented on src/Lucene.Net/Util/ManagedScoreDocArray.cs:
```diff
@@ -0,0 +1,509 @@
+        };
+
+        if (hasFields)
+            segment.Fields = ArrayPool<IComparable[]>.Shared.Rent(size);
```
I'd consider accepting and `ArrayPool<IComparable[]>` in the ctor of this class, then use it here and in the disposal. We should not be making assumptions on such low level. You could also use `ArrayPool<object>` and liberate the caller from understanding what is `IComparable[]` etc.

grisha-kotler replied:
I separated it to `LongArrayPool` and `FieldsArrayPool`.
`FieldsArrayPool` is rarely used.

Scooletz reviewed (changes requested):
I'd pass the ArrayPool from the top and leave the decision of what kind of ArrayPool is it, up to the caller. I'm not experienced in the Lucene though, and its usage in RavenDB, what are all the call sites. I think the majority will be limited to requests and and maybe we could use this as an opportunity to scope the pooling properly.

One more thing. I checked that the project targets .NET 8, which means that `MemoryPool<T>` is available and maybe this should be used as the abstraction here? Right now we don't have any implementation of `MemoryPool<T>` in RavenDB, but if we introduced one, we could use pooling. We could consider building the pool that wraps `ByteStringContext` or similar, memory providing construct and then you'd be sure that no matter how big the query is, we don't pollute server memory and use `long[]` only in this place.

I think this a valid point to consider: Use `MemoryPool<T>` in both cases, pass them from the top, then in RavenDB provide either `.Shared` (is not different from the current) or a custom based on the per-request allocation strategy that already exists.

grisha-kotler committed:
RavenDB-25416 - allow to set a custom ArrayPool

grisha-kotler committed:
RavenDB-25416 - change the BackwardsWriter to a ref struct

grisha-kotler committed:
RavenDB-25416 - explain the architecture

grisha-kotler commented on src/Lucene.Net/Util/ManagedScoreDocArray.cs:
```diff
@@ -0,0 +1,512 @@
+
+                int offset = _indexInSegment * 2;
+                Unsafe.Add(ref intStart, offset) = doc;
+                Unsafe.Add(ref Unsafe.As<int, float>(ref intStart), offset + 1) = score;
```
that is not possible because Lucene doesn't write fields if it was created without them.

grisha-kotler commented on src/Lucene.Net/Search/Searcher.cs:
```diff
@@ -57,9 +57,25 @@ private void  InitBlock()
 		/// <throws>  BooleanQuery.TooManyClauses </throws>
 		public virtual TopFieldDocs Search(Query query, Filter filter, int n, Sort sort, IState state)
 		{
-			return Search(CreateWeight(query, state), filter, n, sort, state);
+			return Search(CreateWeight(query, state), filter, n, sort, fillFields: true, state);
 		}
 		
+        /// <summary>Search implementation with arbitrary sorting.  Finds
+        /// the top <c>n</c> hits for <c>query</c>, applying
+        /// <c>filter</c> if non-null, and sorting the hits by the criteria in
+        /// <c>sort</c>.
+        /// <c>fillFields</c>.
+        /// 
+        /// <p/>NOTE: this does not compute scores by default; use
+        /// <see cref="IndexSearcher.SetDefaultFieldSortScoring(bool,bool)" /> to enable scoring.
+        /// 
```
done

grisha-kotler committed:
RavenDB-25416 - add a comment about the fillFields parameter

grisha-kotler commented on src/Lucene.Net/Util/ManagedScoreDocArray.cs:
```diff
@@ -0,0 +1,512 @@
+    private long[] _currentPacked;
+    public int _currentSegmentUsed;
+    private int _currentSegmentCapacity;
+
```
done

grisha-kotler commented on src/Lucene.Net/Search/TopDocs.cs:
```diff
@@ -16,36 +16,31 @@
-    public class TopDocs
-	{
-		private int _totalHits;
+    [Serializable]
+    public class TopDocs : IDisposable
+    {
+        private int _totalHits;
         private ScoreDoc[] _scoreDocs;
         private float _maxScore;
 
+        public ManagedScoreDocArray ScoreDocArray { get; }
+
```
`ScoreDocArray` includes a finalizer.

Scooletz replied:
Add a comment in the `TopDocs` for future reviews?

grisha-kotler committed:
RavenDB-25416 - address PR comments

grisha-kotler commented on test/Lucene.Net.Test/Search/ManagedScoreDocArrayTests.cs:
```diff
@@ -0,0 +1,404 @@
+    // 1. Core Logic & Growth Phase Tests
+    // ---------------------------------------------------------
+
+    [TestCase(0)]
+    [TestCase(1)]
+    [TestCase(256)]          // End of first segment
+    [TestCase(257)]          // Start of second segment
+    [TestCase(16128)]        // End of Growth Phase (Sum of 256..8192)
+    [TestCase(16129)]        // Start of Stable Phase (First 16k segment)
+    [TestCase(50000)]        // Mid-range
+    [TestCase(128000)]       // Target upper bound
```
done

grisha-kotler committed:
RavenDB-25416 - return the pooled Fields arrays with clearing them because these arrays hold references (IComparable[])

Scooletz commented on src/Lucene.Net/Util/ManagedScoreDocArray.cs:
```diff
@@ -0,0 +1,605 @@
+{
+    public static readonly ManagedScoreDocArray Empty = new();
+
+    public static ArrayPool<long> LongArrayPool = ArrayPool<long>.Shared;
```
As a I understand it, it's a compromise between passing it via a .ctor parameter and having it settable. I think I would prefer to it explicit, as a paramater being passed down from the the top. If it's hard to refactor it this way, we may leave it as is

grisha-kotler replied:
There are currently two places where this is used: RavenDB, where we can pass it through a constructor, and inside Lucene, where things are a bit more constrained. I added it to the parameterless constructor to allow for potential passing through RavenDB.

Scooletz commented on src/Lucene.Net/Util/ManagedScoreDocArray.cs:
```diff
@@ -0,0 +1,512 @@
+
+        foreach (var seg in _segments)
+        {
+            LongArrayPool.Return(seg.PackedDocsAndScores);
```
A managed array not cleared might leak refs @grisha-kotler Could you address it?

Scooletz commented on src/Lucene.Net/Util/ManagedScoreDocArray.cs:
```diff
@@ -0,0 +1,605 @@
+            SwitchToPreviousSegment(doc, score, fields);
+        }
+
+        [MethodImpl(MethodImplOptions.NoInlining)]
```
Why this hint is needed?

grisha-kotler replied:
The `[MethodImpl(MethodImplOptions.NoInlining)]` attribute is needed to keep rarely executed "slow path" code - like allocating new memory segments from cluttering the highly optimized "fast path" methods. By forcing the JIT compiler to keep the fast method small, it ensures the fast path gets cleanly inlined into the main loops without causing CPU instruction cache bloat.

Scooletz replied:
I know what `MethodImplOptions.NoInlining` means. What I meant is that with the advancement of JIT in .NET, 10+, some of these hints are no longer needed as JIT is smart enough to do it for you for free. Let me rephrase the question: have you measured it and showed that the actual inline was happening here, without this attribute applied.

Scooletz commented on src/Lucene.Net/Util/ManagedScoreDocArray.cs:
```diff
@@ -0,0 +1,605 @@
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
```
Why this hint is needed? No inline without it?

grisha-kotler replied:
same here

Scooletz commented on src/Lucene.Net/Util/ManagedScoreDocArray.cs:
```diff
@@ -0,0 +1,605 @@
+            Write(doc, score, fields); // Recursively call Write to hit the fast path
+        }
+
+        private static void ThrowOnEmptyArray() => throw new InvalidOperationException("Cannot write to an empty array");
```
```suggestion
        [DoesNotReturn]
        private static void ThrowOnEmptyArray() => throw new InvalidOperationException("Cannot write to an empty array");
```

grisha-kotler replied:
done

Scooletz commented on src/Lucene.Net/Util/ManagedScoreDocArray.cs:
```diff
@@ -0,0 +1,605 @@
+        }
+
+        private static void ThrowOnEmptyArray() => throw new InvalidOperationException("Cannot write to an empty array");
+        private static void ThrowWriterOutOfRange() => throw new IndexOutOfRangeException("Writer went below index 0");
```
```suggestion
        [DoesNotReturn]
        private static void ThrowWriterOutOfRange() => throw new IndexOutOfRangeException("Writer went below index 0");
```

grisha-kotler replied:
done

Scooletz commented on src/Lucene.Net/Util/ManagedScoreDocArray.cs:
```diff
@@ -0,0 +1,605 @@
+        private static void ThrowWriterOutOfRange() => throw new IndexOutOfRangeException("Writer went below index 0");
+    }
+
+    public ScoreDocReader GetReader(int start) => new(this, start);
```
`ScoreDocReader` is quite a fat struct. Should `out ScoreDocReader reader` be used here maybe? I don't believe creations happens a lot. Leaving it for the author to consider.

grisha-kotler replied:
Although the struct is large, it is only created once per query, meaning this minor one-time copy overhead won't impact the performance of the tight search loops.
Furthermore, modern .NET JIT compilers automatically use a hidden return buffer for large structs anyway, so switching to an `out` parameter wouldn't save CPU cycles but would make the API clunkier.

Scooletz commented on src/Lucene.Net/Util/ManagedScoreDocArray.cs:
```diff
@@ -0,0 +1,605 @@
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public bool Read(out int doc, out float score, out IComparable[] fields)
+        {
+            // Fast path logic duplicated to allow inlining of the "with fields" variant
```
This comment mentions duplication of this special case. I have found no unit test that covers it. If there are none, we should cover it. Potentially, add a test that shows that both `Read` overloads work the same way.

grisha-kotler replied:
Done

Scooletz commented on test/Lucene.Net.Test/Search/TestTopDocsCollector.cs:
```diff
@@ -45,17 +46,17 @@ public MyTopsDocCollector(int size):base(new HitQueue(size, false))
 			}
 			
-			public /*protected internal*/ override TopDocs NewTopDocs(ScoreDoc[] results, int start)
+			public /*protected internal*/ override TopDocs NewTopDocs(ManagedScoreDocArray scoreDocArray, int start)
```
Why there's this comment of `/*protected internal*/`. Remove?

grisha-kotler replied:
That is from the previous version of the code. Removed.

Scooletz commented on src/Lucene.Net/Search/TopDocs.cs:
```diff
@@ -55,18 +50,47 @@ public float MaxScore
+        /// to access the scores without full materialization.</para>
+        /// </remarks>
+        [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Performance", "CA1819:PropertiesShouldNotReturnArrays")]
+        [Obsolete("This property is for testing/debugging ONLY and should not be used in production due to potential memory/performance issues. Use the ScoreDocArray property instead.", error: false)]
```
Thank you for adding this `Obsolete`. I added two comments about how hot this property can get, only to notice that it's debug only 😅

grisha-kotler replied:
There are 654 tests that use `ScoreDocs`.
Changing this would mean that we'll have a lot of changes in this PR :)
Once this is approved, I could change this to using the `ScoreDocArray`.

Scooletz commented on src/Lucene.Net/Search/TopDocs.cs:
```diff
@@ -55,18 +50,47 @@ public float MaxScore
+                if (_scoreDocs == null)
+                {
+                    var reader = ScoreDocArray.GetReader(0);
+                    var list = new List<ScoreDoc>();
+
+                    while (reader.Read(out var doc, out var score))
+                    {
+                        list.Add(new ScoreDoc(doc, score));
+                    }
```
Why not apply this @grisha-kotler ? It looks like the ScoreDocArray has the count?

arekpalinski reviewed (approved):
1. Code changes LGTM. I was unable to spot anything that didn't look correct. I see big potential of this PR.

2. This is a very risky PR. I recommend intensive testing, including:
 - full QA of querying scenarios 
 - long running testing on a variety of internal environments.

 3. My biggest concern is that there might be specific allocation patterns will cause issues that we didn't have so far. It’s hard to predict exactly what these might be but I see them as a potential source of regression. So I suggest a few weeks of testing after merging.

grisha-kotler committed:
RavenDB-25416 - address PR comments

Scooletz commented on src/Lucene.Net/Util/ManagedScoreDocArray.cs:
```diff
@@ -0,0 +1,614 @@
+
+        foreach (var seg in _segments)
+        {
+            _longArrayPool.Return(seg.PackedDocsAndScores);
```
```suggestion
            _longArrayPool.Return(seg.PackedDocsAndScores, clear: false);
```

The clearing is not needed for the long array.

Scooletz replied:
It's false by default. https://github.com/dotnet/runtime/blob/452e9ba0539d628855cef44e08e73795ac1427c8/src/libraries/System.Private.CoreLib/src/System/Buffers/ArrayPool.cs#L97

Scooletz reviewed (approved):
✅  leftover comments from my end, amended / resolved what was hanging.

ppekrol commented on src/Lucene.Net/Util/ManagedScoreDocArray.cs:
```diff
@@ -0,0 +1,614 @@
+        private readonly ManagedScoreDocArray _parent;
+        private ref int _intStart;
+        private ref IComparable[] _fieldsStart;
+        private bool _hasFields;
```
where is this used?

grisha-kotler replied:
removed. i wanted to throw if `_hasFields == false` and someone tries to add a field but that's impossible because we pass the fields only when we create the `ManagedScoreDocArray.cs` with fields.

grisha-kotler committed:
RavenDB-25416 - cleanup

grisha-kotler committed:
RavenDB-25416 - bump lucene version
